docs/production/authentication-methods.md

```diff
@@ -1069,6 +1069,67 @@ configure the JWT secret and algorithm via `JWT_AUTH_KEYS` in
 `/etc/zulip/settings.py`; see the inline comment documentation in that
 file for details.
 
+## Configuring a custom Python wrapper around the `authenticate` mechanism
+
+You can configure a custom function that will work as a wrapper around
+the `authenticate()` method calls of your enabled authentication backend
+classes. Knowledge of [how authentication backends work in Django][django-authenticate-details]
+as well as some familiarity with Zulip's authentication implementation in
+`zproject/backends.py` is required. When in doubt, reach out to us to ask
+for assistance or review of your custom authentication wrapper, especially
+considering that this is a security-sensitive feature and getting it wrong
+may be disastrous to your security model.
+
+In short, when a user is being authenticated to obtain a session
+(and thus this does NOT apply to authentication of API requests
+with a user's API key which use a separate authentication mechanism),
+the `authenticate()` methods of the various authentication backend classes
+in `zproject/backends.py` are called.
+
+You can write custom Python logic that will wrap such `authenticate()`
+calls by specifying in `/etc/zulip/settings.py`:
+
+```
+def custom_auth_wrapper(
+    auth_func, *args, **kwargs
+):
```

Comment:
Is there a reason this can't be defined as:
```
def custom_auth_wrapper(
    auth_func, backend, request, *args, **kwargs,
):
```

Reply:
Because when I was originally testing, I found this doesn't work nicely with social auth backends. Their `authenticate`s are just weird and e.g. can return `HttpResponse`. So the right place to add custom checks in those flows is via the dummy backend. So that's pretty ugly. I added this `important` block, since there should be an explicit note on this and a disclaimer to probably-not-do-it.

![image](https://github.com/zulip/zulip/assets/45007152/faa717e0-66ff-412c-8aee-121d17eac1fe)